src/ossl/fips.rs

```diff
@@ -859,51 +860,83 @@ impl ProviderSignatureCtx {
-                    _ => return Err(CKR_DEVICE_ERROR)?,
-                })
-            }
-        }
-    };
+/* The Openssl FIPS indicator callback is inadequate for easily
+ * accessing individual indicators in the context of a single
+ * operation because it is tied to the general library context,
+ * which can be shared across multiple threads in an application.
+ * Therefore the only way to make this work in a thread safe way
+ * is to use thread local variables */
+thread_local! {
+    static FIPS_INDICATOR: Cell<u32> = Cell::new(0);
 }
 
-check_ossl_fips_indicator!(kdf; KDF; Kdf);
-check_ossl_fips_indicator!(mac; MAC; Mac);
-check_ossl_fips_indicator!(cipher; CIPHER; Cipher);
+unsafe extern "C" fn fips_indicator_callback(
+    _type_: *const ::std::os::raw::c_char,
+    _desc: *const ::std::os::raw::c_char,
+    _params: *const OSSL_PARAM,
+) -> ::std::os::raw::c_int {
+    /* We ignore type, desc, params, for now, and just register
+     * if a change in state occurred.
+     *
+     * We could track individual events in the callback, but
+     * a) it is really hard to know what they are because the
+     *    "type" is an arbitrary string and you need to go and
+     *    find in the specific openssl fips provider sources to
+     *    figure out what it is...
+     * b) it is expensive as it ends up having to do a bunch
+     *    of string compares, and based on that then modify
+     *    some slot in a preallocated vector ...
+     *
+     * Within the context of a thread only one operation at
+     * a time is performed, so, as long as the code correctly
+     * resets the indicator before an operation is started and
+     * immediately checks it at the end, tracking the status in
+     * th operation context, it can get away with tracking
+     * everything in a single per-thread variable and count on
+     * the serial nature of code executing within a thread.
+     *
+     * Note that the callback is called only when the
+     * underlying OpenSSL code believes there was an unapproved
+     * condition. In strict mode the callback is not called and
+     * the underlying function fails directly.
+     */
+
+    /* Set the indicator up, this means there was an unapproved
+     * use. */
+    FIPS_INDICATOR.set(1);
+
+    /* Returning 1, allows OpenSSL to continue the operation.
+     * Unless and until we implement a strict FIPS mode we never
+     * want to cause a failure for an unapproved use, so we just
+     * return all ok, FIPS_INDICATOR will allow us to propagate the
+     * fact that the operation was unapproved by setting PKCS#11
+     * indicators */
+    return 1;
+}
+
+pub fn clear_ossl_fips_indicator() {
+    FIPS_INDICATOR.set(0);
+}
+
+pub fn ossl_fips_indicator_is_set() -> bool {
+    if FIPS_INDICATOR.get() == 0 {
+        return false;
+    }
+    return true;
+}
+
+pub fn fips_approval(fips_approved: &mut Option<bool>, finalize: bool) {
```

Comment:
I find the code now much easier to read than going through the branches of fips in every function. I just had to wrap my head around the arguments as the function call `fips_approval(..., true)` is not completely self-explaining.

My hopefully last nit would be to have a helper function `fips_approval_finalize()` and `fips_approval_check()` that might wrap call this function with the bool parameters inside and use these in the code so it is clear on the first sight what is the function doing with the fips approval without the need to look what arguments names it has.

Reply:
done, and added more functions for intialization too